swarm_dial.go

```diff
@@ -343,23 +343,87 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	if len(goodAddrs) == 0 {
 		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
 	}
-	goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
+
+	///////// Partition and rank addresses
+	// Partition into Fd and non-Fd consuming addresses as we first we want
+	// to try dialling ALL Non-Fd consuming addresses.
 	nonBackoff := false
+	var fdAddrs []ma.Multiaddr
+	var nonFdAddrs []ma.Multiaddr
+
 	for _, a := range goodAddrs {
 		// skip addresses in back-off
 		if !s.backf.Backoff(p, a) {
 			nonBackoff = true
-			goodAddrsChan <- a
+			if s.IsFdConsumingAddr(a) {
+				fdAddrs = append(fdAddrs, a)
+			} else {
+				nonFdAddrs = append(nonFdAddrs, a)
+			}
 		}
 	}
-	close(goodAddrsChan)
 	if !nonBackoff {
 		return nil, ErrDialBackoff
 	}
+
+	/*
+		This slice-to-chan code is temporary, the peerstore can currently provide
+		a channel as an interface for receiving addresses, but more thought
+		needs to be put into the execution. For now, this allows us to use
+		the improved rate limiter, while maintaining the outward behaviour
+		that we previously had (halting a dial when we run out of addrs)
+	*/
+	// addrsToChanFnc transforms a slice of addresses to a closed channel of addresses.
+	addrsToChanFnc := func(goodAddrs []ma.Multiaddr) chan ma.Multiaddr {
+		goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
+		for _, a := range goodAddrs {
+			goodAddrsChan <- a
+		}
+		close(goodAddrsChan)
+		return goodAddrsChan
+	}
+
+	// ranks addresses in descending order of preference for dialing
+	// local addrs > non-relay public addrs > relay-addrs
+	rankAddrsFnc := func(addrs []ma.Multiaddr) []ma.Multiaddr {
+		var localAddrs []ma.Multiaddr
+		var relayAddrs []ma.Multiaddr
+		var others []ma.Multiaddr
```

Comment:
I'd pre-allocate and over-allocate (i.e., with length `len(addrs)`) all of these.

Reply:
Note: as the rules here grow more complicated, it may be simpler to use `sort.Slice()`.

Reply:
@Stebalien Please can you explain why pre-allocating and especially over-allocating would be better than allocating on the go ? What trade-off does that get us ?

Generally speaking, if we have a maximum bound on the size of a slice, why should we always pre-allocate even if that risks over-allocating rather than just allocating on the go ?

Reply:
@Stebalien  Using `sort.Slice now` and got rid of the allocs.